src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Parser.cs

```diff
@@ -0,0 +1,286 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Globalization;
+using System.Linq;
+using System.Threading;
+
+namespace System.Text.RegularExpressions.Generator
+{
+    public partial class RegexGenerator
+    {
+        private const string RegexName = "System.Text.RegularExpressions.Regex";
+        private const string RegexGeneratorAttributeName = "System.Text.RegularExpressions.RegexGeneratorAttribute";
+
+        private static bool IsSyntaxTargetForGeneration(SyntaxNode node) =>
+            // We don't have a semantic model here, so the best we can do is say whether there are any attributes.
+            node is MethodDeclarationSyntax { AttributeLists: { Count: > 0 } };
+
+        private static MethodDeclarationSyntax? GetSemanticTargetForGeneration(GeneratorSyntaxContext context)
+        {
+            var methodDeclarationSyntax = (MethodDeclarationSyntax)context.Node;
+
+            foreach (AttributeListSyntax attributeListSyntax in methodDeclarationSyntax.AttributeLists)
+            {
+                foreach (AttributeSyntax attributeSyntax in attributeListSyntax.Attributes)
+                {
+                    if (context.SemanticModel.GetSymbolInfo(attributeSyntax).Symbol is IMethodSymbol attributeSymbol &&
+                        attributeSymbol.ContainingType.ToDisplayString() == RegexGeneratorAttributeName)
+                    {
+                        return methodDeclarationSyntax;
+                    }
+                }
+            }
+
+            return null;
+        }
+
+        // Returns null if nothing to do, Diagnostic if there's an error to report, or RegexType if the type was analyzed successfully.
+        private static object? GetRegexTypeToEmit(Compilation compilation, MethodDeclarationSyntax methodSyntax, CancellationToken cancellationToken)
+        {
+            // TODO: Use https://github.com/dotnet/runtime/pull/59092
+            INamedTypeSymbol? regexSymbol = compilation.GetTypeByMetadataName(RegexName);
+            INamedTypeSymbol? regexGeneratorAttributeSymbol = compilation.GetTypeByMetadataName(RegexGeneratorAttributeName);
+            if (regexSymbol is null || regexGeneratorAttributeSymbol is null)
+            {
+                // Required types aren't available
+                return null;
+            }
+
+            TypeDeclarationSyntax typeDec = methodSyntax.Parent as TypeDeclarationSyntax;
+            if (typeDec is null)
```

Comment:
Is it not possible to put a Regex attribute in a local method? If someone tries it, will they get a diagnostic?

Reply:
It is possible, and currently there won't be a diagnostic.  With how the Initialize logic was previously structured, this wasn't possible to do well.  Now that it's set up in a way that computes diagnostics earlier in the pipeline and then passes those into Execute, I can add an analysis pass that looks through the whole tree for local functions and sees whether any have the attribute applied.  It will add additional non-trivial cost, though, which is unfortunate when it's being done purely to catch erroneous usage (and you can't write partial locals).

Reply:
> (and you can't write partial locals)

ah right. Maybe we should just leave it as-is then.